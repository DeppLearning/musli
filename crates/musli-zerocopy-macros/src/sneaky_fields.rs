use proc_macro::TokenStream;
use quote::ToTokens;

/// Replace the fields of a struct.
pub(super) fn expand(attr: TokenStream, item: TokenStream) -> TokenStream {
    let mut input = syn::parse_macro_input!(item as syn::Item);
    let ty = syn::parse_macro_input!(attr as syn::Type);

    match &mut input {
        syn::Item::Enum(e) => {
            for v in &mut e.variants {
                match &mut v.fields {
                    syn::Fields::Named(named) => {
                        let extra: syn::FieldsNamed = syn::parse_quote!({ sneaky_field: #ty });
                        named.named.extend(extra.named);
                    }
                    syn::Fields::Unnamed(unnamed) => {
                        let extra: syn::FieldsUnnamed = syn::parse_quote!(( #ty ));
                        unnamed.unnamed.extend(extra.unnamed);
                    }
                    syn::Fields::Unit => {
                        v.fields = syn::Fields::Named(syn::parse_quote!({ sneaky_field: #ty }));
                    }
                }
            }
        }
        syn::Item::Struct(st) => match &mut st.fields {
            syn::Fields::Named(named) => {
                let extra: syn::FieldsNamed = syn::parse_quote!({ sneaky_field: #ty });
                named.named.extend(extra.named);
            }
            syn::Fields::Unnamed(unnamed) => {
                let extra: syn::FieldsUnnamed = syn::parse_quote!(( #ty ));
                unnamed.unnamed.extend(extra.unnamed);
            }
            syn::Fields::Unit => {
                st.fields = syn::Fields::Named(syn::parse_quote!({ sneaky_field: #ty }));
            }
        },
        _ => (),
    }

    input.to_token_stream().into()
}
