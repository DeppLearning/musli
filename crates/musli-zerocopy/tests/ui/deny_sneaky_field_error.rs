//! This test ensures that the derive can correctly see fields which were
//! sneakily introduced by an attribute macro.

use musli_zerocopy::ZeroCopy;
use musli_zerocopy_macros::sneaky_fields;

#[derive(ZeroCopy)]
#[repr(C)]
struct Sneaky(u64);

#[derive(ZeroCopy)]
#[repr(C)]
#[sneaky_fields(Sneaky)]
struct SneakyNamed {
    field: u32,
}

#[sneaky_fields(Sneaky)]
#[derive(ZeroCopy)]
#[repr(C)]
struct SneakyNamed2 {
    field: u32,
}

#[derive(ZeroCopy)]
#[repr(C)]
#[sneaky_fields(Sneaky)]
struct SneakyUnnamed(u32);

#[sneaky_fields(Sneaky)]
#[derive(ZeroCopy)]
#[repr(C)]
struct SneakyUnnamed2(u32);

#[derive(ZeroCopy)]
#[repr(u8)]
#[sneaky_fields(Sneaky)]
enum SneakyEnumNamed {
    Named {
        field: u32,
    },
}

#[derive(ZeroCopy)]
#[repr(u8)]
#[sneaky_fields(Sneaky)]
enum SneakyEnumUnnamed {
    Unnamed(u32)
}

#[derive(ZeroCopy)]
#[repr(u8)]
#[sneaky_fields(Sneaky)]
enum SneakyEnumUnit {
    Unit
}

fn main() {
}
