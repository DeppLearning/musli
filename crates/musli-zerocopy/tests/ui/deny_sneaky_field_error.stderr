error[E0027]: pattern does not mention field `sneaky_field`
  --> tests/ui/deny_sneaky_field_error.rs:14:8
   |
14 |   struct SneakyNamed {
   |  ________^
15 | |     field: u32,
16 | | }
   | |_^ missing field `sneaky_field`
   |
help: include the missing field in the pattern
   |
15 |     field, sneaky_field }
   |          ~~~~~~~~~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
15 |     field, .. }
   |          ~~~~~~

error[E0023]: this pattern has 1 field, but the corresponding tuple struct has 2 fields
  --> tests/ui/deny_sneaky_field_error.rs:25:10
   |
25 | #[derive(ZeroCopy)]
   |          ^^^^^^^^ expected 2 fields, found 1
26 | #[repr(C)]
27 | #[sneaky_fields(Sneaky)]
   |                 ------ tuple struct has 2 fields
28 | struct SneakyUnnamed(u32);
   |        ------------- ---
   |        |
   |        tuple struct defined here
   |
   = note: this error originates in the derive macro `ZeroCopy` (in Nightly builds, run with -Z macro-backtrace for more info)
help: use `_` to explicitly ignore each field
   |
28 | struct SneakyUnnamed(u32, _);
   |                         +++
help: use `..` to ignore all fields
   |
25 | #[derive(..)]
   |          ~~

error[E0027]: pattern does not mention field `sneaky_field`
  --> tests/ui/deny_sneaky_field_error.rs:38:6
   |
38 |   enum SneakyEnumNamed {
   |  ______^
39 | |     Named {
40 | |         field: u32,
41 | |     },
   | |_____^ missing field `sneaky_field`
   |
help: include the missing field in the pattern
   |
40 |         field, sneaky_field },
   |              ~~~~~~~~~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
40 |         field, .. },
   |              ~~~~~~

error[E0023]: this pattern has 1 field, but the corresponding tuple variant has 2 fields
  --> tests/ui/deny_sneaky_field_error.rs:44:10
   |
44 |   #[derive(ZeroCopy)]
   |            ^^^^^^^^ expected 2 fields, found 1
45 |   #[repr(u8)]
46 |   #[sneaky_fields(Sneaky)]
   |                   ------ tuple variant has 2 fields
47 |   enum SneakyEnumUnnamed {
   |  ______-
48 | |     Unnamed(u32)
   | |     ------- ---
   | |_____|_____|
   |       |
   |       tuple variant defined here
   |
   = note: this error originates in the derive macro `ZeroCopy` (in Nightly builds, run with -Z macro-backtrace for more info)
help: use `_` to explicitly ignore each field
   |
48 |     Unnamed(u32, _)
   |                +++
help: use `..` to ignore all fields
   |
44 | #[derive(..)]
   |          ~~

error[E0533]: expected unit struct, unit variant or constant, found struct variant `SneakyEnumUnit::Unit`
  --> tests/ui/deny_sneaky_field_error.rs:54:6
   |
54 |   enum SneakyEnumUnit {
   |  ______^
55 | |     Unit
   | |________^ not a unit struct, unit variant or constant
